cli/fathom_web/extract.py

```diff
@@ -0,0 +1,128 @@
+import base64
+import mimetypes
+import pathlib
+import re
+
+from click import argument, command, Path
+
+
+BASE64_DATA_PATTERN = re.compile(r'(data:(?P<mime>[a-zA-Z0-9]+?/[a-zA-Z0-9\-.+]+?);(\s?charset=utf-8;)?base64,(?P<string>[a-zA-Z0-9+/=]+))')
```

Comment:
Maybe one of your problem pages actually has a "data:blah;blah:base64"-looking string in it that isn't actually a data attr and it's throwing things off? Just a casual idea.

Reply:
I wonder, what made you abandon Beautiful Soup for a grep-based approach? I worry a little about a reduction in precision.

Reply:
I abandoned Beautiful Soup because the page visually changed when doing nothing to the data regardless of which parser I used during input and how I saved the data during output. For example, the code:
```
# image original.png
with open ('267.html') as fp:
    soup = BeautifulSoup(fp)

# image str.png    
with open('new_str.html', 'w') as fp:
    fp.write(str(soup))

# image prettified.png
with open('new_pret.html', 'w') as fp:
    fp.write(soup.prettify())
```
results in the following three pages (in the same order as the code):
![original](https://user-images.githubusercontent.com/5586735/61221458-9096c680-a6e6-11e9-89de-d52903cf5125.png)
In str.png, �'s appear next to the phone number at the top and on the home button
![str](https://user-images.githubusercontent.com/5586735/61221467-92f92080-a6e6-11e9-8b7a-cb0793903d5b.png)
In prettified.png, the home button gets messed up and all of the `|`'s shift
![prettified](https://user-images.githubusercontent.com/5586735/61221468-94c2e400-a6e6-11e9-99c5-6c97b3684715.png)

Reply:
I do think there could be problems with my regex, though it was necessary in the Beautiful Soup version anyway because some of the base64 strings are contained in string attributes of tags that contain a lot of other text. For example in 267.html, one of the base64 strings is here:

```
        <style>
        .top-link-cart{
			background: url(data:image/png;base64,iVBORw...ErkJggg==)  no-repeat scroll left center transparent !important;
			
			 padding:13px 6px 13px 40px !important;
		}
        
        </style>
```
and Beautiful Soup throws everything inside this style tag into a `Tag.string` attribute.

Reply:
Okay, I have found the two changes that solve the remaining problems.

1. Pages with a `<base>` tag need that tag removed.
2. The other two pages (and three of the pages with a `<base>` tag) require setting privacy.file_unique_origin to false in about:config

![cors](https://user-images.githubusercontent.com/5586735/61227349-23892e00-a6f2-11e9-8080-6a27c3bdfe28.png)
https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSRequestNotHttp

For the `<base>` tag, it's still not clear to me why the tag is added.
For the CORS problem, I hope there is a better solution, but I haven't found one yet.

Reply:
Aha. Thanks! I'd be mildly curious to diff the 2 outputted files and see what the changes are.